### PR TITLE
allow user to control if vault decryption is fatal

### LINKED ISSUE #13244: Make vault error type configurable
In some cases diff users might want to use the same play with different access levels, being able to change vault failure to decrypt to a warning or something else allows for this.

### PR DESCRIPTION
**Pull request by bcoca**
##### SUMMARY

fixes #13244 #32687
related #37161 #31141
##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

valut
##### ANSIBLE VERSION

```
2.7
```

### CODE REVIEW

**Comment by AlanCoding** [comment]
+1

File: lib/ansible/parsing/vault/__init__.py [hunk 1]
```diff
@@ -657,7 +657,13 @@ def decrypt(self, vaulttext, filename=None):
...
 
         '''
-        plaintext, vault_id = self.decrypt_and_get_vault_id(vaulttext, filename=filename)
+        try:
+            plaintext, vault_id = self.decrypt_and_get_vault_id(vaulttext, filename=filename)
+        except AnsibleVaultError as e:
+            if C.ERROR_ON_VAULT_FAIL:
+                raise
+            else:
+                display.warning('Vault decryption failed: %s' % to_native(e))
```
Comment (AlanCoding) [inline lib/ansible/parsing/vault/__init__.py]: Now I'm wondering, what will be the value of `plaintext` in this scenario, and what the implications of that will be.

**Comment by AlanCoding** [comment]
I've confirmed my hunch.

You can see that both cases still fail. It doesn't really accomplish a pass-through if the setting is set to `False`, as one would expect. There is a different message emitted in both cases, so your conditional really is intercepting the missing secrets case, but there appears to be more cases that also need to be handled.

I would still like to have this, I would be happy to incorporate use of this constant into my other PR.

##### false case
```
$ ANSIBLE_ERROR_ON_VAULT_FAIL=false ansible-playbook -i scripts/vault/awx_redumper.py debugging/hostvars_print.yml

PLAY [all] ****************************************************************************************************************************************************************************

TASK [debug] **************************************************************************************************************************************************************************
 [WARNING]: Vault decryption failed: Attempting to decrypt but no vault secrets found

fatal: [foobar]: FAILED! => {"msg": "An unhandled exception occurred while templating '{{ hostvars[inventory_hostname] }}'. Error was a <type 'exceptions.UnboundLocalError'>, original message: local variable 'plaintext' referenced before assignment"}
	to retry, use: --limit @/Users/alancoding/Documents/repos/ansible-inventory-file-examples/debugging/hostvars_print.retry

PLAY RECAP ****************************************************************************************************************************************************************************
foobar                     : ok=0    changed=0    unreachable=0    failed=1   

```

##### true case
```
$ ANSIBLE_ERROR_ON_VAULT_FAIL=true ansible-playbook -i scripts/vault/awx_redumper.py debugging/hostvars_print.yml

PLAY [all] ****************************************************************************************************************************************************************************

TASK [debug] **************************************************************************************************************************************************************************
fatal: [foobar]: FAILED! => {"msg": "An unhandled exception occurred while templating '{{ hostvars[inventory_hostname] }}'. Error was a <class 'ansible.parsing.vault.AnsibleVaultError'>, original message: Attempting to decrypt but no vault secrets found"}
	to retry, use: --limit @/Users/alancoding/Documents/repos/ansible-inventory-file-examples/debugging/hostvars_print.retry

PLAY RECAP ****************************************************************************************************************************************************************************
foobar                     : ok=0    changed=0    unreachable=0    failed=1  
```

**Comment by bcoca** [comment]
might need to change then, ignore_missing_secrets vs ignore_failed_decrypt, probably need the 2nd toggle a few steps up.

**Comment by bcoca** [comment]
@AlanCoding i updated to both cover all types of errors and make the distinction between missing/mismatched secrets vs other general vault errors.

File: lib/ansible/config/base.yml [hunk 2]
```diff
@@ -1244,6 +1244,22 @@ ERROR_ON_MISSING_HANDLER:
   ini:
   - {key: error_on_missing_handler, section: defaults}
   type: boolean
+ERROR_ON_VAULT_FAIL:
+  name: Error on vault decryption failure
+  default: True
+  description: "Toggle to speicfy if 'non-secret' related vault decryption failures are errors or warnings."
```
Comment (alikins) [inline lib/ansible/config/base.yml]: s/speicfy/specify

File: lib/ansible/config/base.yml [hunk 3]
```diff
@@ -1244,6 +1244,22 @@ ERROR_ON_MISSING_HANDLER:
...
+  env: [{name: ANSIBLE_ERROR_ON_VAULT_FAIL}]
+  ini:
+  - {key: error_on_fail, section: vault}
+  type: boolean
+ERROR_ON_VAULT_SECRET:
+  name: Error on missing/mismatched vault secrets
+  default: True
+  description: "Toggle to speicfy if vault decryption caused by missing or mismatch secrets to be errors or warnings."
```
Comment (alikins) [inline lib/ansible/config/base.yml]: s/speicfy/specify

**Comment by AlanCoding** [comment]
There's a much better solution to the particular use case I was interested in via 

https://github.com/ansible/ansible/issues/37536
https://github.com/ansible/ansible/issues/37531

The particular behavior of "don't decrypt vault values" can reasonably be accomplished with a simple flag on `ansible-inventory`

**Pull request closed without merging**